system/src/system_power_manager.cpp

```diff
@@ -552,52 +557,188 @@ void PowerManager::handleStateChange(battery_state_t from, battery_state_t to, b
 #endif // defined(DEBUG_BUILD)
 }
 
-battery_state_t PowerManager::handlePossibleFault(battery_state_t from, battery_state_t to) {
-  if (to == BATTERY_STATE_CHARGED || to == BATTERY_STATE_CHARGING) {
-    system_tick_t m = millis();
-    if (m - possibleFaultTimestamp_ > DEFAULT_FAULT_WINDOW) {
-      possibleFaultTimestamp_ = m;
-      possibleFaultCounter_ = 0;
-      faultSecondaryCounter_ = 0;
+void PowerManager::batteryStateForwardingTo(battery_state_t targetState, bool count) {
+  if (targetState == BATTERY_STATE_NOT_CHARGING && g_batteryState != BATTERY_STATE_NOT_CHARGING) {
+    if (count) {
+      notChargingDebounceCount_++;
     } else {
-      possibleFaultCounter_++;
-      if (possibleFaultCounter_ >= DEFAULT_FAULT_COUNT_THRESHOLD &&
-         (faultSuppressed_ == 0 || (m - faultSuppressed_ >= DEFAULT_FAULT_SUPPRESSION_PERIOD))) {
-        if (faultSecondaryCounter_ > 0) {
-          faultSecondaryCounter_ = 0;
-          return BATTERY_STATE_DISCONNECTED;
-        } else {
-          PMIC power;
-          power.setRechargeThreshold(PMIC_FAULT_RECHARGE_THRESHOLD);
-          possibleFaultCounter_ = 0;
-          faultSecondaryCounter_ = 1;
-          possibleFaultTimestamp_ = millis();
-        }
-      }
+      notChargingDebounceCount_ = 0;
+    }
+    chargedFaultCount_ = 0;
+    chargingDebounceCount_ = 0;
+    lastChargedTimeStamp_ = 0;
+    battMonitorPeriod_ = BATTERY_STATE_CHANGE_CHECK_PERIOD;
+    DBG_PWR("notChargingDebounceCount_: %d", notChargingDebounceCount_);
+    if (notChargingDebounceCount_ >= BATTERY_NOT_CHARGING_DEBOUNCE_COUNT) {
+      confirmBatteryState(g_batteryState, BATTERY_STATE_NOT_CHARGING);
+    }
+  } else if (targetState == BATTERY_STATE_CHARGING && g_batteryState != BATTERY_STATE_CHARGING) {
+    if (count) {
+      chargingDebounceCount_++;
+    } else {
+      chargingDebounceCount_ = 0;
+    }
+    chargedFaultCount_ = 0;
+    notChargingDebounceCount_ = 0;
+    lastChargedTimeStamp_ = 0;
+    battMonitorPeriod_ = BATTERY_STATE_CHANGE_CHECK_PERIOD;
+    DBG_PWR("chargingDebounceCount_: %d", chargingDebounceCount_);
+    if (chargingDebounceCount_ >= BATTERY_CHARGING_DEBOUNCE_COUNT) {
+      confirmBatteryState(g_batteryState, BATTERY_STATE_CHARGING);
+    }
+  } else if (targetState == BATTERY_STATE_CHARGED) {
```

Comment:
Do we count transitions from CHARGED to CHARGED as well?

Reply:
It does, right? whichever the current state is, as long as the target state is charged we count it. That's how we count the charged event to detect disconnected state.